File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    // Write your first then lower case function here
    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream()
                .filter(predicate)
                .map(String::toLowerCase)
                .findFirst();
    }

    // Write your say function here
    
    private String phrase;

    private Exercises(String phrase) {
        this.phrase = phrase;
    }

    public static Exercises say(String... words) {
    return new Exercises(String.join(" ", words).trim());
    }

    public Exercises and(String word) {
        phrase += " " + word;
        return this;
    }

    public String phrase() {
        return phrase.trim();
    }



    // Write your line count function here
    
    public static int meaningfulLineCount(String filePath) throws IOException {
        int meaningfulLinesCount = 0;

        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (!line.isEmpty() && !line.startsWith("#")) {
                    meaningfulLinesCount++;
                }
            }
        } catch (IOException e) {
            throw new IOException("No such file: " + filePath, e);
        }

        return meaningfulLinesCount;
    }
}

// Write your Quaternion record class here
record Quaternion(double a, double b, double c, double d) {

    static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    static final Quaternion I = new Quaternion(0, 1, 0, 0);
    static final Quaternion J = new Quaternion(0, 0, 1, 0);
    static final Quaternion K = new Quaternion(0, 0, 0, 1);

    Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    Quaternion plus(Quaternion q) {
        return new Quaternion(a + q.a, b + q.b, c + q.c, d + q.d);
    }

    Quaternion times(Quaternion q) {
        return new Quaternion(
            a * q.a - b * q.b - c * q.c - d * q.d,
            a * q.b + b * q.a + c * q.d - d * q.c,
            a * q.c - b * q.d + c * q.a + d * q.b,
            a * q.d + b * q.c - c * q.b + d * q.a
        );
    }

    Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        appendTerm(sb, a, "");
        appendTerm(sb, b, "i");
        appendTerm(sb, c, "j");
        appendTerm(sb, d, "k");
        return sb.length() == 0 ? "0" : sb.toString();
    }

    private void appendTerm(StringBuilder sb, double coeff, String symbol) {
        if (coeff == 0) return;
        if (sb.length() > 0 && coeff > 0) sb.append("+");
        sb.append(coeff == 1 || coeff == -1 ? (coeff < 0 ? "-" : "") + symbol : coeff + symbol);
    }
}




// Write your BinarySearchTree sealed interface and its implementations here
sealed interface BinarySearchTree permits Empty, Node {
    int size();
    boolean contains(String s);
    BinarySearchTree insert(String s);
}

final class Empty implements BinarySearchTree {
    public int size() {
        return 0; 
    }

    public boolean contains(String s) {
        return false; 
    }

    public BinarySearchTree insert(String s) {
        return new Node(s); 
    }

    public String toString() {
        return "()"; 
    }
}

final class Node implements BinarySearchTree {
    String data;
    BinarySearchTree left;
    BinarySearchTree right;

    // Constructor for creating a new Node
    Node(String data) {
        this.data = data;
        this.left = new Empty();
        this.right = new Empty();
    }

    // New constructor for creating a Node with left and right children
    Node(String data, BinarySearchTree left, BinarySearchTree right) {
        this.data = data;
        this.left = left;
        this.right = right;
    }

    public int size() {
        return 1 + left.size() + right.size();
    }

    public BinarySearchTree insert(String s) {
        if (s.compareTo(this.data) < 0) {
            return new Node(this.data, left.insert(s), right); 
        } else if (s.compareTo(this.data) > 0) {
            return new Node(this.data, left, right.insert(s)); 
        } else {
            return this; // No changes made if the value already exists
        }
    }

    public boolean contains(String s) {
        if (this.data.equals(s)) {
            return true;
        } else if (s.compareTo(this.data) < 0) {
            return left.contains(s); 
        } else {
            return right.contains(s);
        }
    }

    public String toString() {
        return ("(" + left.toString() + this.data + right.toString() + ")").replace("()", ""); 
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// Write your first then lower case function here
fun firstThenLowerCase(words: List<String>, predicate: (String) -> Boolean): String? {
    return words.firstOrNull(predicate)?.lowercase()
}

// Write your say function here
class Say(private val sentence: String = "") {
    val phrase: String
        get() = sentence.trim()

    fun and(nextWord: String): Say {
        return Say("$sentence $nextWord")
    }
}

fun say(word: String = ""): Say {
    return Say(word)
}

// Write your meaningfulLineCount function here
fun meaningfulLineCount(filePath: String): Long {
    val inputFile = java.io.File(filePath)
    if (!inputFile.exists()) {
        throw IOException("No such file")
    }

    return inputFile.readLines()
        .filterNot { it.isBlank() || it.startsWith("//") }
        .count().toLong()
}

// Write your Quaternion data class here
data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    operator fun plus(q: Quaternion): Quaternion {
        return Quaternion(a + q.a, b + q.b, c + q.c, d + q.d)
    }

    operator fun times(q: Quaternion): Quaternion {
        return Quaternion(
            a * q.a - b * q.b - c * q.c - d * q.d,
            a * q.b + b * q.a + c * q.d - d * q.c,
            a * q.c - b * q.d + c * q.a + d * q.b,
            a * q.d + b * q.c - c * q.b + d * q.a
        )
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)

    override fun toString(): String {
        return buildString {
            if (a != 0.0) append(a)
            if (b != 0.0) append("${if (b > 0 && isNotEmpty()) "+" else ""}${if (b == 1.0) "" else b}i")
            if (c != 0.0) append("${if (c > 0 && isNotEmpty()) "+" else ""}${if (c == 1.0) "" else c}j")
            if (d != 0.0) append("${if (d > 0 && isNotEmpty()) "+" else ""}${if (d == 1.0) "" else d}k")
            if (isEmpty()) append("0")
        }
    }
}

// Write your Binary Search Tree interface and implementing classes here
sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree
    override fun toString(): String

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }
}

data class Node(val value: String, val left: BinarySearchTree = BinarySearchTree.Empty, val right: BinarySearchTree = BinarySearchTree.Empty) : BinarySearchTree {
    override fun size(): Int = 1 + left.size() + right.size()

    override fun contains(value: String): Boolean {
        return when {
            value == this.value -> true
            value < this.value -> left.contains(value)
            else -> right.contains(value)
        }
    }

    override fun insert(value: String): BinarySearchTree {
        return when {
            value < this.value -> Node(this.value, left.insert(value), right)
            value > this.value -> Node(this.value, left, right.insert(value))
            else -> this
        }
    }

    override fun toString(): String {
        val leftStr = if (left is BinarySearchTree.Empty) "" else "($left)"
        val rightStr = if (right is BinarySearchTree.Empty) "" else "($right)"
        return "$leftStr$value$rightStr"
    }
}
